server/src/unifyfs_request_manager.c

```diff
@@ -816,121 +816,151 @@ int rm_cmd_exit(reqmgr_thrd_t* thrd_ctrl)
-               unifyfs_keys[i]->fid, unifyfs_keys[i]->offset,
-               unifyfs_vals[i]->len, unifyfs_vals[i]->app_id,
-               unifyfs_vals[i]->rank);
+               key->fid, key->offset,
+               val->len, val->app_id, val->rank);
 
+        /* MDHIM needs to know the byte size of each key and value */
         unifyfs_key_lens[i] = sizeof(unifyfs_key_t);
         unifyfs_val_lens[i] = sizeof(unifyfs_val_t);
     }
 
+    /* batch insert file extent key/values into MDHIM */
     ret = unifyfs_set_file_extents((int)extent_num_entries,
                                    unifyfs_keys, unifyfs_key_lens,
                                    unifyfs_vals, unifyfs_val_lens);
     if (ret != UNIFYFS_SUCCESS) {
-        // TODO: need proper error handling
+        /* TODO: need proper error handling */
         LOGERR("unifyfs_set_file_extents() failed");
         goto rm_cmd_fsync_exit;
     }
 
-    // file attributes
-    attr_num_entries = *(size_t*)
-        (app_config->shm_superblocks[client_side_id]
-         + app_config->fmeta_offset);
+    /* get number of file attribute values client has for us,
+     * stored as a size_t value in fmeta region of shared memory */
+    size_t attr_num_entries = *(size_t*)(fmeta);
 
-    /*
-     * file attributes are stored in the superblock shared memory
-     * created by the client
-     */
-    unifyfs_file_attr_t* attr_payload = (unifyfs_file_attr_t*)
-        (app_config->shm_superblocks[client_side_id]
-         + app_config->fmeta_offset + page_sz);
-
-    // allocate storage for values
-    // TODO: possibly get this from memory pool
-    fattr_keys = alloc_attr_key_array(attr_num_entries);
-    fattr_vals = calloc(attr_num_entries, sizeof(unifyfs_file_attr_t*));
+    /* file attributes are stored in the superblock shared memory
+     * created by the client */
+    char* ptr_fattr = fmeta + page_sz;
+    unifyfs_file_attr_t* attr_payload = (unifyfs_file_attr_t*)(ptr_fattr);
+
+    /* allocate storage for file attribute key/values */
+    /* TODO: possibly get this from memory pool */
+    fattr_keys     = alloc_attr_key_array(attr_num_entries);
+    fattr_vals     = calloc(attr_num_entries, sizeof(unifyfs_file_attr_t*));
     fattr_key_lens = calloc(attr_num_entries, sizeof(int));
     fattr_val_lens = calloc(attr_num_entries, sizeof(int));
     if ((NULL == fattr_keys) ||
         (NULL == fattr_vals) ||
         (NULL == fattr_key_lens) ||
         (NULL == fattr_val_lens)) {
+        LOGERR("failed to allocate memory for file attributes");
         ret = (int)UNIFYFS_ERROR_NOMEM;
         goto rm_cmd_fsync_exit;
     }
 
+    /* create file attribute key/values for insertion into MDHIM */
     for (i = 0; i < attr_num_entries; i++) {
+        /* for a key, we use the global file id */
         *fattr_keys[i] = attr_payload[i].gfid;
+
+        /* for the value, we use address of file attribute? */
+        /* TODO: this means each server must have a local copy for each file,
```

Comment:
MDHIM doesn't store the pointer as the val, it copies the data at the given pointer with the given size. There is a bug here though, since the ptr is to a unifyfs_file_attr_t struct, but we give the size of fattr_val_t struct.

Reply:
Oh, yeah I just realized that's what MDHIM was doing.

Let me check the difference on structs.

Reply:
AFAICT, fattr_val_t is no longer used for anything, and should be removed.

Reply:
Yes, I concur.  Doing that now.